src/vector/_pytree.py

```diff
@@ -0,0 +1,176 @@
+"""PyTree operations for vector objects.
+
+This module defines how vector objects are handled within optree.
+See https://blog.scientific-python.org/pytrees/ for the rationale for these functions.
+"""
+
+from __future__ import annotations
+
+from functools import partial
+from typing import TYPE_CHECKING, Any
+
+import numpy
+
+if TYPE_CHECKING:
+    from optree.pytree import ReexportedPyTreeModule
+
+from vector._methods import (
+    Vector2D,
+    Vector3D,
+    Vector4D,
+)
+from vector.backends.numpy import (
+    MomentumNumpy2D,
+    MomentumNumpy3D,
+    MomentumNumpy4D,
+    VectorNumpy,
+    VectorNumpy2D,
+    VectorNumpy3D,
+    VectorNumpy4D,
+)
+from vector.backends.object import (
+    MomentumObject2D,
+    MomentumObject3D,
+    MomentumObject4D,
+    VectorObject2D,
+    VectorObject3D,
+    VectorObject4D,
+)
+
+Children = tuple[Any, ...]
+MetaData = tuple[type, ...]
+
+
+def _flatten2D(v: Vector2D) -> tuple[Children, MetaData]:
+    children = v.azimuthal.elements
+    metadata = type(v), type(v.azimuthal)
+    return children, metadata
+
+
+def _unflatten2D(metadata: MetaData, children: Children) -> Vector2D:
+    backend, azimuthal = metadata
+    return backend(azimuthal=azimuthal(*children))
+
+
+def _flatten3D(v: Vector3D) -> tuple[Children, MetaData]:
+    children = v.azimuthal.elements, v.longitudinal.elements
+    metadata = type(v), type(v.azimuthal), type(v.longitudinal)
+    return children, metadata
+
+
+def _unflatten3D(metadata: MetaData, children: Children) -> Vector3D:
+    coords_azimuthal, coords_longitudinal = children
+    backend, azimuthal, longitudinal = metadata
+    return backend(
+        azimuthal=azimuthal(*coords_azimuthal),
+        longitudinal=longitudinal(*coords_longitudinal),
+    )
+
+
+def _flatten4D(v: Vector4D) -> tuple[Children, MetaData]:
+    children = (
+        v.azimuthal.elements,
+        v.longitudinal.elements,
+        v.temporal.elements,
+    )
+    metadata = type(v), type(v.azimuthal), type(v.longitudinal), type(v.temporal)
+    return children, metadata
+
+
+def _unflatten4D(metadata: MetaData, children: Children) -> Vector4D:
+    coords_azimuthal, coords_longitudinal, coords_temporal = children
+    backend, azimuthal, longitudinal, temporal = metadata
+    return backend(
+        azimuthal=azimuthal(*coords_azimuthal),
+        longitudinal=longitudinal(*coords_longitudinal),
+        temporal=temporal(*coords_temporal),
+    )
+
+
+def _flattenAoSdata(v: VectorNumpy) -> tuple[Children, tuple[type, numpy.dtype]]:
+    assert v.dtype.fields is not None
+    field_dtypes = [dt for dt, *_ in v.dtype.fields.values()]
+    target_dtype = field_dtypes[0]
+    if not all(fd == target_dtype for fd in field_dtypes):
+        raise ValueError("All fields must have the same dtype to flatten")
+    array = numpy.array(v).view(target_dtype)
```

Comment:
We manually flatten the numpy struct dtype here and include the necessary information to revert it. For safety we require all fields to have the same dtype. This is necessary to fully flatten the array to a 1d vector